### ISSUE TITLE
Should gradients of logsumexp([-inf, -inf]) be 0s instead NaNs?

### ISSUE THREAD

**Issue by kho** [google/jax#6811]
Consider this snippet

```python
import jax
import jax.numpy as jnp
def f(x):
  return jax.nn.logsumexp(jnp.stack([-jnp.inf, -jnp.inf + x]))
jax.grad(f)(1.)
```

The computation inside `f` is essentially what happens inside the dynamic program of the CTC loss and similar losses that calculates a shortest distance over the log semiring. There, `-jnp.inf` is really "log(0)", and we can see that `(0 + 0 * exp(x))`'s gradient w.r.t. `x` should indeed by 0.

Right now, the gradients are NaN instead because the gradients of `logsumexp([-inf, -inf])` are NaNs.

**Comment by jakevdp** [comment]
Thanks for the suggestion - I think that makes sense. I'll note that this is only relevant in the case where *every* entry in the array is negative infinity; if there are finite entries the result is as expected:
```python
grad(logsumexp)(jnp.array([-jnp.inf, 1]))
```
so perhaps the could be accomplished with a conditional in the function rather than a full custom JVP.

**Comment by peterroelants** [comment]
I'be been running into a similar issue, and was wondering if anyone has a code snippet as workaround?

**Comment by kho** [comment]
I have been using the following implementation with a custom VJP:

```
@functools.partial(jax.custom_vjp, nondiff_argnums=(1,))
def _logsumexp(a, axis=None):
  return _logsumexp_fwd(a, axis)[0]


def _logsumexp_fwd(a, axis):
  c = jnp.max(a, axis=axis, keepdims=True)
  safe = jnp.isfinite(c)
  c = jnp.where(safe, c, 0)
  e = jnp.exp(a - c)
  z = jnp.sum(e, axis=axis, keepdims=True)
  r = jnp.squeeze(c, axis=axis) + jnp.log(jnp.squeeze(z, axis=axis))
  return r, (e, z)


def _logsumexp_bwd(axis, res, g):
  e, z = res
  safe = z != 0
  z = jnp.where(safe, z, 1)
  if axis is not None:
    g = jnp.expand_dims(g, axis=axis)
  return (g * jnp.where(safe, e / z, 0),)


_logsumexp.defvjp(_logsumexp_fwd, _logsumexp_bwd)
```

I also gave some thought @jakevdp's suggestion and came up with the following:

```
def logsumexp(a, axis=None):
  c = jnp.max(a, axis=axis, keepdims=True)
  safe = jnp.isfinite(c)
  c = jnp.where(safe, c, 0)
  e = jnp.exp(a - c)
  z = jnp.sum(e, axis=axis)
  safe = z != 0
  log_z = jnp.where(safe, jnp.log(jnp.where(safe, z, 1)), -jnp.inf)
  r = jnp.squeeze(c, axis=axis) + log_z
  return r
```

The custom VJP solution uses 3 where ops in total when computing the gradients but the one without needs 6, which feels a bit less than ideal to me.

**Comment by ihh** [comment]
@kho thanks for this. I believe there is a typo in `_logsumexp_fwd`. The line where `c` is reassigned should read

~~~
  c = jnp.where(safe, 0, c)
~~~

as opposed to

~~~
  c = jnp.where(safe, c, 0)
~~~

**Comment by kho** [comment]
That's not a typo. Try this test case: `_logsumexp_fwd(jnp.array([jnp.inf, 1]), axis=0)[0]`. It should return inf not NaN.

**Comment by ihh** [comment]
Oh yes thank you for clarifying, sorry about that
